### PR TITLE
Add quad triplet optimization example

### PR DESCRIPTION
**Pull request by cemitch99**
Added an example of a quadrupole triplet focusing system, illustrating matching of a beam to small spot size at focus (target values of beta_x = beta_y and alpha_x = alpha_y = 0).

This example will be modified to illustrate the use of optimization.

- [x] Add example input with nominal (near-optimal) parameters
- [x] Update examples/CmakeLists.txt
- [x] Update documentation
- [x] Add plot script
- [x] Modify Python script to allow optimization of quad strengths and/or spacing

### CODE REVIEW

File: examples/optimize_triplet/README.rst [hunk 1]
```diff
@@ -0,0 +1,56 @@
+.. _examples-triplet:
+
+Optimized Triplet
+===================
+
+Optimization of focusing parameters for a quadrupole triplet.
+A 1 GeV electron beam is strongly focused from lattice initial parameters:
```
Comment (n01r) [inline examples/optimize_triplet/README.rst]: The script initializes a 2 GeV electron beam - should it initialize a 1 GeV beam or does the value here need to be updated?

File: examples/optimize_triplet/input_triplet.in [hunk 2]
```diff
@@ -0,0 +1,57 @@
+###############################################################################
+# Particle Beam(s)
+###############################################################################
+beam.npart = 10000
+beam.units = static
+beam.kin_energy = 2.0e3
```
Comment (n01r) [inline examples/optimize_triplet/input_triplet.in]: 2 GeV initialized here - should that read 1 GeV?

File: examples/optimize_triplet/run_triplet.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+sim.space_charge = False
+# sim.diagnostics = False  # benchmarking
+sim.slice_step_diagnostics = True
+
+# domain decomposition & space charge mesh
+sim.init_grids()
+
+# load a 2 GeV electron beam with an initial
+# unnormalized rms emittance of 5 nm
+kin_energy_MeV = 2.0e3  # reference energy
```
Comment (n01r) [inline examples/optimize_triplet/run_triplet.py]: as above, 2 GeV here - should it read 1 GeV?

File: examples/optimize_triplet/analysis_triplet.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+print("")
+print("Final Beam:")
+sigx, sigy, sigt, emittance_x, emittance_y, emittance_t = get_moments(final)
+print(f"  sigx={sigx:e} sigy={sigy:e} sigt={sigt:e}")
+print(
+    f"  emittance_x={emittance_x:e} emittance_y={emittance_y:e} emittance_t={emittance_t:e}"
+)
+
+atol = 0.0  # ignored
+rtol = 2.0 * num_particles**-0.5  # from random sampling of a smooth distribution
```
Comment (n01r) [inline examples/optimize_triplet/analysis_triplet.py:83]: the value for `rtol` is smaller for the final beam than for the initial beam ... could that lead to the initial beam passing the check while the final one fails it?

**Review by n01r — COMMENTED** [review]
Thank you, @cemitch99! Looks good 🙂 
I marked some tiny details with comments.
Will the last bullet point about the actual optimization also be done in this PR or on a follow-up?

File: examples/optimize_triplet/analysis_triplet.py (hunk 4 above)
Comment (cemitch99) [reply examples/optimize_triplet/analysis_triplet.py:83]: Thanks.  This was copied from a previous script--both values of rtol are now set to the same number.

File: examples/optimize_triplet/input_triplet.in (hunk 2 above)
Comment (cemitch99) [reply examples/optimize_triplet/input_triplet.in]: The discrepancy was in the README.  The correct value is 2 GeV, and the README has been updated.

File: examples/optimize_triplet/run_triplet.py (hunk 3 above)
Comment (cemitch99) [reply examples/optimize_triplet/run_triplet.py]: See comment above.

**Comment by cemitch99** [comment]
Here is an illustration of the beta functions.
[Optimized_Triplet.pdf](https://github.com/ECP-WarpX/impactx/files/14452852/Optimized_Triplet.pdf)

File: examples/optimize_triplet/README.rst [hunk 5]
```diff
@@ -0,0 +1,56 @@
+.. _examples-triplet:
+
+Optimized Triplet
+===================
```
Comment (ax3l) [inline examples/optimize_triplet/README.rst]:
```suggestion
Optimized Triplet
=================
```

**Review by ax3l — COMMENTED** [review]
Thank you very much!

Let's add the visualization of the beta functions as a visualization script (as we do in the FODO example)?

**Comment by cemitch99** [comment]
Good idea.  Working on the visualization script now.

**Comment by ax3l** [comment]
Thank you so much. I'll try to add the optimization part this week :)

**Comment by ax3l** [comment]
cs_twiss_beta

File: examples/optimize_triplet/README.rst [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+You can run the following script to visualize the beam evolution over time:
+
+.. dropdown:: Script ``plot_triplet.py``
+
+   .. literalinclude:: plot_triplet.py
+      :language: python3
+      :caption: You can copy this file from ``examples/fodo/plot_triplet.py``.
+
+.. figure:: https://private-user-images.githubusercontent.com/1353258/309984511-ba3ca6a0-e73e-4c69-9e7e-b60685433a6b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2MTM4NjQsIm5iZiI6MTcwOTYxMzU2NCwicGF0aCI6Ii8xMzUzMjU4LzMwOTk4NDUxMS1iYTNjYTZhMC1lNzNlLTRjNjktOWU3ZS1iNjA2ODU0MzNhNmIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDVUMDQzOTI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODhkZWMyNjkxYjQwZjgzNGNiYzU5N2UxMmM0ZGY5YTRmYTkzZGJkNmY5Mzg4NzAzMTFlMTEyMGMyNGFjZjU1OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.yfNmqJohzemrw5LQum6zeqfGPxhAi9VEBXH_ieEALro
```
Comment (ax3l) [inline examples/optimize_triplet/README.rst]: I need to put the image on gist.

File: examples/optimize_triplet/README.rst [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+
+Run
+---
+
+This example can be run as a **Python** script: ``python3 run_triplet.py``.
+
+.. literalinclude:: run_triplet.py
+  :language: python3
+  :caption: You can copy this file from ``examples/optimize_triplet/run_triplet.py``.
```
Comment (ax3l) [inline examples/optimize_triplet/README.rst]: I need to add tabs here for `scipy.optimize` and `xopt`

**Review by ax3l — APPROVED** [review]

**Pull request merged**